### ISSUE TITLE
Model forward outputs to text/image similarity score

### ISSUE THREAD

**Issue by paulcjh** [lucidrains/x-clip#8]
Any insight on how to take the image/text embeddings (or nominal model forward output) to achieve a simple similarity score as done in the huggingface implementation? [HF example here](https://huggingface.co/docs/transformers/model_doc/clip#usage)

In the original paper I see the dot products of the image/text encoder outputs were used, but here I was having troubles with the dimensions on the outputs.

**Comment by paulcjh** [comment]
@rom1504 That's what I'm roughly trying, but both encoder outputs are from transformers and have 3 dimensions.
Also, which cosine function in torch are you referring to here? The standard one takes only one arg (normal cos), which one are you referring to?

**Comment by rom1504** [comment]
cosine is the dot product between the normalized image embeddings and the normalized text embeddings

**Comment by rom1504** [comment]
but anyway, did you train this using some thousands of gpu hours already?
otherwise you're not going to get meaningful embeddings by using randomly initialized models

**Comment by paulcjh** [comment]
@rom1504 Here is a small snippet of my code with the tensor sizes, the cosine function only works when I use a fixed dimension from the transformer output (dim 1 in either model output). But I'm sure this is incorrect unless I'm missing something?
<img width="1114" alt="Screenshot 2022-05-09 at 01 19 42" src="https://user-images.githubusercontent.com/47833405/167321866-4e4cd326-6fae-4059-b658-bb9ea88d9f16.png">

**Comment by paulcjh** [comment]
Dimensions work when simply doing the standard model output, not with the `return_encodings=True`. Look correct to you, @lucidrains does this look right to you? Seems obvious in hindsight.
<img width="1125" alt="Screenshot 2022-05-09 at 11 57 38" src="https://user-images.githubusercontent.com/47833405/167396507-bc8ac498-a0ff-41a3-bdab-8cfcf7953685.png">

**Comment by lucidrains** [comment]
@paulcjh Hi Paul, I believe you should be able to get the similarity scores simply by ensuring `return_loss` is not set to `True`? https://github.com/lucidrains/x-clip/blob/main/x_clip/x_clip.py#L599 could you confirm that this is or is not working?

**Comment by paulcjh** [comment]
@lucidrains hey, just checked now and I'm getting a tuple of two tensors, each is `[256, 4]` in shape. Looks like the 256 is matching my input text length (`text_seq_length`) and I'm inputting 4 samples. This sound correct to you, I can drop all of my model args here if needed

**Comment by lucidrains** [comment]
@paulcjh hey Paul, i believe that must mean you still have `return-encodings` turned on

if you paste the whole script (which shouldn't be too long) then i can probably point out what needs to be corrected

**Comment by paulcjh** [comment]
@lucidrains awesome thanks! Omitted some of the unnecessary items:
```
tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
tokenizer.pad_token = tokenizer.eos_token
model = CLIP(
    dim_text = 512,
    dim_image = 512,
    dim_latent = 512,
    num_text_tokens = 50257,                # Number of tokens in GPT2 Tokenizer
    text_enc_depth = 6,
    text_seq_len = 256,
    text_heads = 8,
    text_pad_id = tokenizer.pad_token_id,
    visual_enc_depth = 6,
    visual_image_size = 64,
    visual_patch_size = 32,
    visual_heads = 8,
    use_all_token_embeds = True,            # whether to use fine-grained contrastive learning (FILIP)
    decoupled_contrastive_learning = True,  # use decoupled contrastive learning (DCL) objective function, removing positive pairs from the denominator of the InfoNCE loss (CLOOB + DCL)
    extra_latent_projection = True,         # whether to use separate projections for text-to-image vs image-to-text comparisons (CLOOB)
    use_visual_ssl = True,                  # whether to do self supervised learning on iages
    visual_ssl_type = 'simclr',             # can be either 'simclr' or 'simsiam', depending on using DeCLIP or SLIP
    use_mlm = False,                        # use masked language learning (MLM) on text (DeCLIP)
    text_ssl_loss_weight = 0.05,            # weight for text MLM loss
    image_ssl_loss_weight = 0.05            # weight for image self-supervised learning loss
)
...
data = next(iter(training_loader))
sample_inputs, sample_labels = dataset.batch_to_tensor(data)

# View sample image
sample_index = 18
sample_image = sample_inputs[sample_index]
sample_label = sample_labels[sample_index]
sample_image

from io import BytesIO
import requests
from skimage import io
plt.figure()

processed_image = torch.Tensor(torch.permute(sample_image,[1,2,0]))
processed_image = (processed_image + 1) * 0.5
print(f'Caption: {dataset.tokenizer.decode(sample_label, skip_special_tokens=True)}')
plt.imshow(processed_image)
plt.show()  

# Run through clip
test_classes = ["tree", "dog", "ocean", "sea"]
sample_caption_prefix = "this is a picture of a "

input_tokens = dataset.tokenizer([sample_caption_prefix + _class for _class in test_classes],
                                        padding="max_length",
                                        max_length=dataset.input_seq_length, 
                                        return_tensors="pt")['input_ids'].to(device)

model_output = model(input_tokens, torch.unsqueeze(sample_inputs[sample_index], 0).to(device), return_encodings=False)

print(model_output[1].shape)
print(model_output[0].shape)
torch.nn.Softmax(dim=0)(torch.nn.CosineSimilarity(dim=0)(model_output[0], model_output[1]))
```
Thanks again!

**Comment by lucidrains** [comment]
@paulcjh Hi Paul, could you remove the matplotlib instance too for clarity? and the tokenizer as well and just substitute with a mock tensor of longs. So one thing off the bat is that `use_all_token_embeds` should be set to `False` (that is a relative new paper that proposes fine grained interactions between text and image encodings https://arxiv.org/abs/2111.07783)

**Comment by paulcjh** [comment]
@lucidrains Hey Phil, just updated my code and made an obvious error when playing with the output in leaving the slicing of the 0 & 1 dimension, I see now that the output is [4, 256, 4] which I need to get to the [4] similarity vect.

Also, noted the comment about `use_all_token_embeds` good thing I hadn't committed too much time to training. Will change that.
